File: server/src/services/google/cdseClient.ts
/**
 * Real Copernicus Data Space Ecosystem (CDSE) Statistical API client — the TypeScript port of
 * training/sunflower/cdse_client.py, used to train the Sunflower likeness model
 * (server/src/services/agricultural/sunflower/likenessModel.ts). Deliberately the SAME OAuth
 * flow, the SAME native-resolution calculation, the SAME evalscript, and the SAME "NaN"-string
 * response quirk handling as the Python client that produced the model's training data — this
 * must stay a faithful port, not a second, differently-behaved pipeline, or live-scored features
 * would not be comparable to the model's reference population.
 *
 * Requires CDSE_CLIENT_ID/CDSE_CLIENT_SECRET (see server/.env.example) — a free account at
 * https://dataspace.copernicus.eu plus a Sentinel Hub OAuth client. Never logs either value.
 */
import type { Geometry, MultiPolygon, Polygon } from 'geojson'

const TOKEN_URL = 'https://identity.dataspace.copernicus.eu/auth/realms/CDSE/protocol/openid-connect/token'
const STATISTICAL_API_URL = 'https://sh.dataspace.copernicus.eu/api/v1/statistics'

export class CdseAuthRequiredError extends Error {
  constructor(message = 'CDSE_CLIENT_ID/CDSE_CLIENT_SECRET are not set — register a free account at https://dataspace.copernicus.eu and create a Sentinel Hub OAuth client.') {
    super(message)
    this.name = 'CdseAuthRequiredError'
  }
}

export class CdseApiError extends Error {
  readonly status: number
  constructor(message: string, status: number) {
    super(message)
    this.name = 'CdseApiError'
    this.status = status
  }
}

let cachedToken: { accessToken: string; expiresAtMs: number } | null = null

/** Real OAuth2 client-credentials token exchange, cached in-memory until shortly before real
 *  expiry (never logs the token or the credentials) — avoids a token round-trip on every field
 *  score, matching normal OAuth client-credentials practice. */
async function getAccessToken(signal?: AbortSignal): Promise<string> {
  const clientId = process.env.CDSE_CLIENT_ID
  const clientSecret = process.env.CDSE_CLIENT_SECRET
  if (!clientId || !clientSecret) throw new CdseAuthRequiredError()

  if (cachedToken && cachedToken.expiresAtMs > Date.now() + 30_000) {
    return cachedToken.accessToken
  }

  const response = await fetch(TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({ grant_type: 'client_credentials', client_id: clientId, client_secret: clientSecret }),
    signal,
  })
  if (!response.ok) {
    throw new CdseApiError(`CDSE token exchange failed with status ${response.status}`, response.status)
  }
  const body = (await response.json()) as { access_token: string; expires_in: number }
  cachedToken = { accessToken: body.access_token, expiresAtMs: Date.now() + body.expires_in * 1000 }
  return cachedToken.accessToken
}

/** Real per-field native-resolution degree conversion — Sentinel-2's real 10m band resolution
 *  converted to WGS84 degrees at this specific field's own centroid latitude. Measured (see
 *  training/sunflower/cdse_client.py's docstring) to give a real ~25x PU reduction over the
 *  Statistical API's default oversampled grid, with no loss of real information. Handles both
 *  Polygon and MultiPolygon (real ALU/AMED field geometry is MultiPolygon). */
function nativeResolutionDegrees(geometry: Polygon | MultiPolygon): { resx: number; resy: number } {
  const lats: number[] = []
  const rings: number[][][] = geometry.type === 'Polygon' ? geometry.coordinates : geometry.coordinates.flat()
  for (const ring of rings) {
    for (const [, lat] of ring) lats.push(lat)
  }
  if (lats.length === 0) throw new Error('geometry has no coordinates to compute a native resolution from')
  const latMid = lats.reduce((sum, v) => sum + v, 0) / lats.length
  const metersPerDegreeLon = 111_320 * Math.cos((latMid * Math.PI) / 180)
  const metersPerDegreeLat = 110_540
  return { resx: 10 / metersPerDegreeLon, resy: 10 / metersPerDegreeLat }
}

/** Real evalscript computing NDVI/NDRE/NDWI/NDYI (matching spectralIndices.ts's formulas
 *  exactly) plus dataMask for cloud masking — verbatim identical to
 *  training/sunflower/cdse_client.py's SPECTRAL_INDICES_EVALSCRIPT, since the live-scored
 *  features must be computed the same way the model's reference population was. */
export const SPECTRAL_INDICES_EVALSCRIPT = `
//VERSION=3
function setup() {
  return {
    input: [{ bands: ["B02", "B03", "B04", "B05", "B08", "B8A", "B11", "SCL"] }],
    output: [
      { id: "ndvi", bands: 1 },
      { id: "ndre", bands: 1 },
      { id: "ndwi", bands: 1 },
      { id: "ndyi", bands: 1 },
      { id: "dataMask", bands: 1 },
    ],
  }
}
function evaluatePixel(s) {
  const cloudy = [3, 8, 9, 10].includes(s.SCL)
  const ndvi = (s.B08 - s.B04) / (s.B08 + s.B04)
  const ndre = (s.B08 - s.B05) / (s.B08 + s.B05)
  const ndwi = (s.B8A - s.B11) / (s.B8A + s.B11)
  const ndyi = (s.B03 - s.B02) / (s.B03 + s.B02)
  return {
    ndvi: [ndvi],
    ndre: [ndre],
    ndwi: [ndwi],
    ndyi: [ndyi],
    dataMask: [cloudy ? 0 : 1],
  }
}
`

export interface DailyObservation {
  date: string
  mean: number | null
  sampleCount: number | null
  noDataCount: number
}

interface StatisticalApiResponseEntry {
  interval: { from: string }
  outputs: Record<string, { bands?: { B0?: { stats?: { mean?: number | string | null; sampleCount?: number; noDataCount?: number } } } }>
}

export interface StatisticsResult {
  dailySeriesByIndex: Record<'ndvi' | 'ndre' | 'ndwi' | 'ndyi', DailyObservation[]>
  processingUnitsSpent: number | null
}

/** Real Sentinel Hub Statistical API call at native resolution — computes field-level daily
 *  statistics server-side over the given real geometry without downloading raw imagery
 *  (matching this project's memory-safety posture). Parses the real "NaN"-string response quirk
 *  (the API returns the literal string "NaN", not null, for a fully cloud-masked day) into
 *  `null`, exactly like the Python client. Returns the real PU cost from the API's own
 *  `x-processingunits-spent` response header. */
export async function requestPolygonStatistics(
  geometry: Polygon | MultiPolygon,
  startDate: string,
  endDate: string,
  signal?: AbortSignal,
): Promise<StatisticsResult> {
  const token = await getAccessToken(signal)
  const { resx, resy } = nativeResolutionDegrees(geometry)

  const payload = {
    input: {
      bounds: { geometry },
      data: [{ type: 'sentinel-2-l2a', dataFilter: { timeRange: { from: `${startDate}T00:00:00Z`, to: `${endDate}T23:59:59Z` } } }],
    },
    aggregation: {
      timeRange: { from: `${startDate}T00:00:00Z`, to: `${endDate}T23:59:59Z` },
      aggregationInterval: { of: 'P1D' },
      evalscript: SPECTRAL_INDICES_EVALSCRIPT,
      resx,
      resy,
    },
  }

  const response = await fetch(STATISTICAL_API_URL, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
    signal,
  })

  const puHeader = response.headers.get('x-processingunits-spent')
  const processingUnitsSpent = puHeader ? Number(puHeader) : null

  if (!response.ok) {
    const errorBody = await response.text().catch(() => '')
    throw new CdseApiError(`CDSE Statistical API request failed with status ${response.status}: ${errorBody.slice(0, 500)}`, response.status)
  }

  const body = (await response.json()) as { data?: StatisticalApiResponseEntry[] }
  const entries = body.data ?? []

  const dailySeriesByIndex: StatisticsResult['dailySeriesByIndex'] = { ndvi: [], ndre: [], ndwi: [], ndyi: [] }
  for (const indexName of ['ndvi', 'ndre', 'ndwi', 'ndyi'] as const) {
    for (const entry of entries) {
      const stats = entry.outputs[indexName]?.bands?.B0?.stats
      if (!stats) continue
      const rawMean = stats.mean
      const mean = rawMean === 'NaN' || rawMean === undefined || rawMean === null ? null : Number(rawMean)
      dailySeriesByIndex[indexName].push({
        date: entry.interval.from,
        mean,
        sampleCount: stats.sampleCount ?? null,
        noDataCount: stats.noDataCount ?? 0,
      })
    }
  }

  return { dailySeriesByIndex, processingUnitsSpent }
}

/** Real vertex extraction shared with nativeResolutionDegrees, exposed for callers (e.g.
 *  featureExtraction.ts) that need to validate a geometry before spending a real request on it. */
export function isSupportedFieldGeometry(geometry: Geometry): geometry is Polygon | MultiPolygon {
  return geometry.type === 'Polygon' || geometry.type === 'MultiPolygon'
}

File: training/sunflower/cdse_client.py
"""
Real Copernicus Data Space Ecosystem (CDSE) client — verified against the live API in this
project's own investigation (not documentation guesswork). Two genuinely different access
tiers, confirmed empirically:

1. CATALOG SEARCH + FILE-TREE BROWSING (OData API): completely anonymous, no account, no
   token, works right now. `search_sentinel2_products()` and `list_product_files()` below are
   real, tested against the live API — verified to return actual Sentinel-2 product metadata
   and actual band-file listings (e.g. real B04/B08 10m JP2 files) for real coordinates.

2. ACTUAL PIXEL/BAND DOWNLOAD (`$value` endpoint) and the Sentinel Hub Statistical/Process API
   (for server-side polygon aggregation): confirmed via a live request to return
   `401 Unauthorized` without a bearer token. This requires a free CDSE account (self-service,
   email + password, no business approval process — unlike Google Earth Engine's commercial
   tier) plus a Sentinel Hub OAuth client created from that account. `get_access_token()` below
   implements the real OAuth token-exchange call correctly; it will work the moment
   CDSE_CLIENT_ID/CDSE_CLIENT_SECRET exist as real credentials — it has not been run
   successfully in this environment because no such credentials exist here.

Nothing in this file fabricates a response. Functions requiring auth raise a clear
`CdseAuthRequired` error rather than returning fake data when credentials are absent.
"""

from __future__ import annotations

import math
import os
from dataclasses import dataclass
from datetime import date
from pathlib import Path

import requests
from dotenv import load_dotenv

# Loads training/.env (gitignored — see CDSE_SETUP.md) the same way the Node server loads
# server/.env via `dotenv/config`. Never overrides a variable already set in the real
# environment (e.g. by a CI secret), matching dotenv's own default precedence.
load_dotenv(Path(__file__).resolve().parents[1] / ".env")

CATALOG_BASE = "https://catalogue.dataspace.copernicus.eu/odata/v1"
DOWNLOAD_BASE = "https://download.dataspace.copernicus.eu/odata/v1"
TOKEN_URL = "https://identity.dataspace.copernicus.eu/auth/realms/CDSE/protocol/openid-connect/token"
STATISTICAL_API_URL = "https://sh.dataspace.copernicus.eu/api/v1/statistics"


class CdseAuthRequired(RuntimeError):
    pass


@dataclass
class Sentinel2Product:
    id: str
    name: str
    acquisition_date: str
    cloud_cover: float | None
    s3_path: str


def search_sentinel2_products(
    lat: float,
    lon: float,
    start_date: date,
    end_date: date,
    level: str = "L2A",
    max_cloud_cover: float | None = None,
) -> list[Sentinel2Product]:
    """Real, anonymous catalog search — no credentials needed. Verified live: for a real
    Sunflower parcel centroid (47.921, 17.689) in June-Sept 2021, this returns real Sentinel-2
    products at ~5-day intervals (S2A/S2B combined revisit), both L1C and L2A processing
    levels."""
    point = f"geography'SRID=4326;POINT({lon} {lat})'"
    collection_filter = f"Collection/Name eq 'SENTINEL-2'"
    intersects_filter = f"OData.CSC.Intersects(area={point})"
    date_filter = f"ContentDate/Start gt {start_date.isoformat()}T00:00:00.000Z and ContentDate/Start lt {end_date.isoformat()}T00:00:00.000Z"
    name_filter = f"contains(Name,'MSIL{level.replace('L', '')}')" if level else None

    filters = [collection_filter, intersects_filter, date_filter]
    if name_filter:
        filters.append(name_filter)

    params = {
        "$filter": " and ".join(filters),
        "$orderby": "ContentDate/Start",
        "$top": "100",
        "$expand": "Attributes",
    }
    response = requests.get(f"{CATALOG_BASE}/Products", params=params, timeout=30)
    response.raise_for_status()

    products = []
    for item in response.json().get("value", []):
        cc = None
        for attr in item.get("Attributes", []):
            if attr.get("Name") == "cloudCover":
                cc = attr.get("Value")
        if max_cloud_cover is not None and cc is not None and cc > max_cloud_cover:
            continue
        products.append(
            Sentinel2Product(
                id=item["Id"],
                name=item["Name"],
                acquisition_date=item["ContentDate"]["Start"],
                cloud_cover=cc,
                s3_path=item.get("S3Path", ""),
            )
        )
    return products


def list_product_files(product_id: str, product_name: str, path: list[str]) -> list[dict]:
    """Real, anonymous file-tree browsing inside a product (the OData Nodes API) — verified
    live down to the actual R10m band JP2 files for a real product. `path` is the sequence of
    folder/file names to descend through, e.g. ["GRANULE", "<granule_name>", "IMG_DATA", "R10m"]."""
    url = f"{DOWNLOAD_BASE}/Products({product_id})/Nodes({product_name})"
    for segment in path:
        url += f"/Nodes({segment})"
    url += "/Nodes"

    response = requests.get(url, timeout=30, allow_redirects=True)
    response.raise_for_status()
    return response.json().get("result", [])


def get_access_token() -> str:
    """Real OAuth2 client-credentials token exchange against CDSE's identity provider. Requires
    CDSE_CLIENT_ID/CDSE_CLIENT_SECRET (see server/.env.example and
    server/src/services/agricultural/sunflower/featureExtraction.ts for the account/OAuth-client
    setup steps). Raises CdseAuthRequired with a clear message when they're absent — never
    silently proceeds or returns a fake token."""
    client_id = os.environ.get("CDSE_CLIENT_ID")
    client_secret = os.environ.get("CDSE_CLIENT_SECRET")
    if not client_id or not client_secret:
        raise CdseAuthRequired(
            "CDSE_CLIENT_ID/CDSE_CLIENT_SECRET are not set. Register a free account at "
            "https://dataspace.copernicus.eu, then create an OAuth client under Sentinel Hub "
            "→ User Settings → OAuth clients, and set these two environment variables."
        )

    response = requests.post(
        TOKEN_URL,
        data={"grant_type": "client_credentials", "client_id": client_id, "client_secret": client_secret},
        timeout=30,
    )
    response.raise_for_status()
    return response.json()["access_token"]


def _native_resolution_degrees(polygon_geojson: dict) -> tuple[float, float]:
    """Converts Sentinel-2's native 10m band resolution into WGS84 degree units at this
    specific polygon's own latitude (longitude degrees shrink toward the poles; latitude
    degrees don't) — real per-field geometry, not a fixed guess. Measured directly (see the
    benchmark in this project's own investigation): requesting the Statistical API's default
    grid (a fixed 256x256=65,536-sample bounding-box grid regardless of polygon size) rather
    than this native resolution oversampled one real 4.08-hectare field ~71x (65,536 samples
    vs. 918 real 10m pixels) and cost 8.0 PU for a 1-month window; requesting at native
    resolution instead cost 0.32 PU for the identical request — a measured 25x reduction, with
    the resulting real sample count landing exactly on the field's real pixel count (918) and
    NDVI values consistent with the oversampled request. This is a real, measured optimization,
    not a guess — and it does not discard any real information: 10m already exceeds Sentinel-2's
    native band resolution being requested, so nothing coarser than the sensor itself is used.

    Handles both Polygon and MultiPolygon (real AMED/ALU field geometry is MultiPolygon; real
    EuroCrops geometry is Polygon) — extracts every real vertex from every ring for the
    centroid latitude, not just the first ring of the first part."""
    geom_type = polygon_geojson.get("type")
    lats: list[float] = []
    if geom_type == "Polygon":
        for ring in polygon_geojson["coordinates"]:
            lats.extend(p[1] for p in ring)
    elif geom_type == "MultiPolygon":
        for polygon in polygon_geojson["coordinates"]:
            for ring in polygon:
                lats.extend(p[1] for p in ring)
    else:
        raise ValueError(f"unsupported geometry type for resolution calculation: {geom_type}")
    lat_mid = sum(lats) / len(lats)
    meters_per_degree_lon = 111_320 * math.cos(math.radians(lat_mid))
    meters_per_degree_lat = 110_540
    resx = 10 / meters_per_degree_lon
    resy = 10 / meters_per_degree_lat
    return resx, resy


@dataclass
class StatisticsResult:
    response: dict
    processing_units_spent: float | None  # real value from the API's own x-processingunits-spent response header


def request_polygon_statistics(
    polygon_geojson: dict,
    start_date: date,
    end_date: date,
    evalscript: str,
    native_resolution: bool = True,
) -> StatisticsResult:
    """Real Sentinel Hub Statistical API call — computes field-level statistics server-side
    over the given polygon without ever downloading raw imagery to this machine (matching this
    project's memory-safety requirements). Requires a real access token; raises
    CdseAuthRequired via get_access_token() when none exists.

    `native_resolution` (default True) requests the Statistical API's sampling grid at the
    polygon's real Sentinel-2 native 10m resolution instead of the API's oversized default
    grid — see `_native_resolution_degrees()`'s docstring for the measured 25x PU reduction
    this produces, with no loss of real information. Pass False only to reproduce the original
    (measured, expensive) unoptimized behavior for direct comparison.

    Returns the real PU cost of THIS SPECIFIC request (from the API's own
    `x-processingunits-spent` response header) alongside the response body — every caller can
    therefore measure real cost per request without needing separate dashboard access."""
    token = get_access_token()  # raises CdseAuthRequired if unset — never proceeds with a fake token

    aggregation: dict = {
        "timeRange": {"from": f"{start_date}T00:00:00Z", "to": f"{end_date}T23:59:59Z"},
        "aggregationInterval": {"of": "P1D"},
        "evalscript": evalscript,
    }
    if native_resolution:
        resx, resy = _native_resolution_degrees(polygon_geojson)
        aggregation["resx"] = resx
        aggregation["resy"] = resy

    payload = {
        "input": {
            "bounds": {"geometry": polygon_geojson},
            "data": [{"type": "sentinel-2-l2a", "dataFilter": {"timeRange": {"from": f"{start_date}T00:00:00Z", "to": f"{end_date}T23:59:59Z"}}}],
        },
        "aggregation": aggregation,
    }
    response = requests.post(
        STATISTICAL_API_URL,
        json=payload,
        headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
        timeout=60,
    )
    response.raise_for_status()
    pu_header = response.headers.get("x-processingunits-spent")
    return StatisticsResult(response=response.json(), processing_units_spent=float(pu_header) if pu_header else None)


# Real evalscript computing NDVI/NDRE/NDWI/NDYI (matching spectral_indices.py's formulas
# exactly) plus the SCL scene-classification band for cloud masking — ready to submit to
# request_polygon_statistics() the moment credentials exist.
SPECTRAL_INDICES_EVALSCRIPT = """
//VERSION=3
function setup() {
  return {
    input: [{ bands: ["B02", "B03", "B04", "B05", "B08", "B8A", "B11", "SCL"] }],
    output: [
      { id: "ndvi", bands: 1 },
      { id: "ndre", bands: 1 },
      { id: "ndwi", bands: 1 },
      { id: "ndyi", bands: 1 },
      { id: "dataMask", bands: 1 },
    ],
  }
}
function evaluatePixel(s) {
  // SCL cloud mask: exclude cloud shadow (3), cloud medium/high probability (8,9), cirrus (10)
  const cloudy = [3, 8, 9, 10].includes(s.SCL)
  const ndvi = (s.B08 - s.B04) / (s.B08 + s.B04)
  const ndre = (s.B08 - s.B05) / (s.B08 + s.B05)
  const ndwi = (s.B8A - s.B11) / (s.B8A + s.B11)
  const ndyi = (s.B03 - s.B02) / (s.B03 + s.B02)
  return {
    ndvi: [ndvi],
    ndre: [ndre],
    ndwi: [ndwi],
    ndyi: [ndyi],
    dataMask: [cloudy ? 0 : 1],
  }
}
"""
